server/src/main/scala/actors/daily/UpdatesSupervisor.scala

```diff
@@ -0,0 +1,132 @@
+}
+
+class UpdatesSupervisor[A, B <: WithTimeAccessor](now: () => SDateLike,
+                                                  terminals: List[Terminal],
+                                                  updatesActorFactory: (Terminal, SDateLike) => Props) extends Actor {
+  val log: Logger = LoggerFactory.getLogger(getClass)
+  implicit val ex: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
+  implicit val mat: ActorMaterializer = ActorMaterializer.create(context)
+  implicit val timeout: Timeout = new Timeout(15 seconds)
+  val cancellableTick: Cancellable = context.system.scheduler.schedule(10 seconds, 10 seconds, self, PurgeExpired)
+  val killActor: ActorRef = context.system.actorOf(Props(new RequestAndTerminateActor()))
+
+  var streamingUpdateActors: Map[(Terminal, MillisSinceEpoch), ActorRef] = Map[(Terminal, MillisSinceEpoch), ActorRef]()
+  var lastRequests: Map[(Terminal, MillisSinceEpoch), MillisSinceEpoch] = Map[(Terminal, MillisSinceEpoch), MillisSinceEpoch]()
+
+  override def postStop(): Unit = {
+    log.warn("Actor stopped. Cancelling scheduled tick")
+    cancellableTick.cancel()
+    super.postStop()
+  }
+
+  def startUpdatesStream(terminal: Terminal,
+                         day: SDateLike): Unit = streamingUpdateActors.get((terminal, day.millisSinceEpoch)) match {
+    case Some(_) => Unit
+    case None =>
+      log.info(s"Starting supervised updates stream for $terminal / ${day.toISODateOnly}")
+      val actor = context.system.actorOf(updatesActorFactory(terminal, day))
+      streamingUpdateActors = streamingUpdateActors + ((terminal, day.millisSinceEpoch) -> actor)
+      lastRequests = lastRequests + ((terminal, day.millisSinceEpoch) -> now().millisSinceEpoch)
+  }
+
+  override def receive: Receive = {
+    case PurgeExpired =>
+      log.info("Received PurgeExpired")
+      val expiredToRemove = lastRequests.collect {
+        case (tm, lastRequest) if now().millisSinceEpoch - lastRequest > MilliTimes.oneMinuteMillis =>
+          log.info(s"Shutting down streaming updates for ${tm._1}/${SDate(tm._2).toISODateOnly}")
+          streamingUpdateActors.get(tm).foreach(_ ! PoisonPill)
+          tm
+      }
+      streamingUpdateActors = streamingUpdateActors -- expiredToRemove
+      lastRequests = lastRequests -- expiredToRemove
+
+    case StartUpdatesStream(terminal, day) =>
+      startUpdatesStream(terminal, day)
+      sender() ! Ack
+
+    case GetUpdatesSince(sinceMillis, fromMillis, toMillis) =>
+      val replyTo = sender()
+      val terminalDays = terminalDaysForPeriod(fromMillis, toMillis)
+
+      terminalsAndDaysUpdatesSource(terminalDays, sinceMillis)
+        .runWith(Sink.fold(MinutesContainer.empty[A, B])(_ ++ _))
+        .onComplete {
+          case Success(container) => replyTo ! container
+          case Failure(t) =>
+            log.error("Failed to get updates", t)
+            replyTo ! MinutesContainer.empty[A, B]
```

Comment:
is sending an empty list here going to disguise the error for the recipient of the message? Should this be an option rather, or have a failure case? It's a problem we already have elsewhere, but maybe this is a place we can address it.

Reply:
yeah good call

Reply:
I've pushed the error handling up to the end point function so we return a 503 response in the case of any error. I've also removed the `Either[PortStateError, Option[PortState]]` in favour of just `Option[PortState]`.